File: leetcode/234-palindrome-linked-list.py
# https://leetcode.com/problems/palindrome-linked-list/ (top-interview-questions)

from typing import Optional

# Definition for singly-linked list.
class ListNode:
	def __init__(self, val=0, next=None):
		self.val = val
		self.next = next

class Solution:
	def isPalindrome(self, head: Optional[ListNode]) -> bool:
		slow, fast = head, head

		# find the middle node
		while fast and fast.next:
			fast = fast.next.next
			slow = slow.next

		# reverse the second half
		palindrome = None
		while slow:
			next = slow.next
			slow.next = palindrome
			palindrome = slow
			slow = next

		# check if the values of the first half and
		# the second (after reversing it) are equal.
		while palindrome:
			if head.val != palindrome.val:
				return False
			head = head.next
			palindrome = palindrome.next
		return True

File: leetcode/234-palindrome-linked-list.js
// https://leetcode.com/problems/palindrome-linked-list/ (top-interview-questions)


// Definition for singly-linked list.
function ListNode(val, next) {
  this.val = (val===undefined ? 0 : val)
  this.next = (next===undefined ? null : next)
}

/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function(head) {

  // get the middle node
  let slow = head;
  let fast = head;
  while (fast && fast.next) {
    slow = slow.next;
    fast = fast.next?.next;
  }

  // reverse the linkedlist's second half 
  prev = null;
  while (slow) {
    const next = slow.next;
    slow.next = prev;
    prev = slow;
    slow = next;
  }

  // palindrom checking...
  while (prev) {
    if (prev.val !== head.val) {
      return false;
    }
    prev = prev.next;
    head = head.next;
  }

  return true;
};
